sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+	struct virtqueue *ipc_psn_vq;
+
+	/* audio data in both directions */
+	struct virtqueue *data_vq;
+
+	/* position update work */
+	struct work_struct posn_update_work;
+
+	/* current pending cmd message */
+	struct snd_sof_ipc_msg *msg;
+
+	/* current and pending notification */
+	struct snd_sof_ipc_msg *not;
+	struct sof_ipc_stream_posn posn;
+
+	struct sof_vfe_ipc_tplg_resp tplg;
+
+	struct completion completion;
+
+	/* A shared capture / playback virtual queue data buffer */
+	union {
+		struct dsp_sof_data_req data_req;
+		struct dsp_sof_data_resp data_resp;
+	};
+
+	/* Headers, used as a playback response or capture request */
+	union {
+		u8 hdr_req[HDR_SIZE_REQ];
+		u8 hdr_resp[HDR_SIZE_RESP];
+	};
+};
+
+/*
+ * IPC Firmware ready.
+ */
+static int sof_vfe_fw_ready(struct snd_sof_dev *sdev, u32 msg_id)
+{
+	return 0;
+};
+
+/* used to send IPC to BE */
+static int sof_vfe_send_msg(struct snd_sof_dev *sdev,
+			    struct snd_sof_ipc_msg *msg)
+{
+	struct sof_vfe *vfe = sdev->pdata->vfe;
+	int ret;
+	struct scatterlist sg_out, sg_in, *sgs[] = {&sg_out, &sg_in};
+	size_t msg_size = msg->msg_size;
+	void *msg_data = msg->msg_data;
+
+	switch (msg->header) {
+	case SOF_IPC_GLB_STREAM_MSG | SOF_IPC_STREAM_TRIG_START:
+		sg_init_one(&sg_in, &vfe->posn,
+			    sizeof(struct sof_ipc_stream_posn));
+		ret = virtqueue_add_inbuf(vfe->ipc_psn_vq,
+					  &sg_in, 1, &vfe->posn, GFP_KERNEL);
+		if (ret < 0) {
+			dev_err(sdev->dev, "%s(): failed %d to add a buffer\n",
+				__func__, ret);
+			return ret;
+		}
+
+		virtqueue_kick(vfe->ipc_psn_vq);
+
+		break;
+	}
+
+	sg_init_one(&sg_out, msg_data, msg_size);
+	sg_init_one(&sg_in, msg->reply_data, msg->reply_size);
+
+	ret = virtqueue_add_sgs(vfe->ipc_cmd_vq, sgs, 1, 1, msg_data,
+				GFP_ATOMIC);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: could not send IPC %d\n", ret);
+
+	vfe->msg = msg;
+
+	virtqueue_kick(vfe->ipc_cmd_vq);
+
+	return ret;
+}
+
+/* Handle playback or capture data */
+static void sof_vfe_handle_data(struct virtqueue *vq)
+{
+	struct sof_vfe *vfe = vq->vdev->priv;
+
+	complete(&vfe->completion);
+}
+
+/* Send the IPC message completed. This means vBE has received the cmd */
+static void sof_vfe_cmd_tx_done(struct virtqueue *vq)
+{
+	struct sof_vfe *vfe = vq->vdev->priv;
+	DEFINE_SPINLOCK(lock);
+
+	do {
+		struct snd_sof_ipc_msg *msg = vfe->msg;
+		struct sof_ipc_reply *reply = msg->reply_data;
+		unsigned int len;
+
+		virtqueue_disable_cb(vq);
+
+		spin_lock(&lock);
+		/*
+		 * virtqueue_get_buf() returns the "token" that was provided to
+		 * virtqueue_add_*() functions
+		 */
+		while (virtqueue_get_buf(vq, &len)) {
+			msg->reply_error = reply->error;
+
+			/* Firmware panic? */
+			if (msg->reply_error == -ENODEV)
+				vfe->sdev->ipc->disable_ipc_tx = true;
+
+			msg->ipc_complete = true;
+			wake_up(&msg->waitq);
+		}
+		spin_unlock(&lock);
+	} while (!virtqueue_enable_cb(vq));
+}
+
+static void sof_vfe_posn_update(struct work_struct *work)
+{
```

Comment:
I would state here that this is the "copy" version and is suseptable to latency.

Reply:
done